### PR TITLE
incr.comp.: Make DepNode's std::fmt::Debug implementation useful again.

### PR DESCRIPTION
**Pull request by michaelwoerister**
With #42537 a regular `DepNode` only contains an opaque hash as its identifier. In most cases, this hash is actually a `DefPathHash` and we can reconstruct the `DefId` it came from via a table lookup --- and then use that to print something intelligible for debug outputs. For cases where we cannot reconstruct information from the DepNode's hash, this PR will cache a string representation of the `DepNode` in a side-table. This string is later used for debug outputs.

r? @nikomatsakis

### CODE REVIEW

**Review by nikomatsakis — APPROVED** [review]

File: src/librustc/dep_graph/dep_node.rs [hunk 1]
```diff
@@ -457,6 +517,62 @@ impl<'a, 'gcx: 'tcx + 'a, 'tcx: 'a> DepNodeParams<'a, 'gcx, 'tcx> for (DefId,) {
...
+}
+
+impl<'a, 'gcx: 'tcx + 'a, 'tcx: 'a> DepNodeParams<'a, 'gcx, 'tcx> for (DefId, DefId) {
+    const CAN_RECONSTRUCT_QUERY_KEY: bool = false;
+
+    fn to_fingerprint(&self, tcx: TyCtxt) -> Fingerprint {
+        let (def_id_0, def_id_1) = *self;
+
+        let def_path_hash_0 = tcx.def_path_hash(def_id_0);
+        let def_path_hash_1 = tcx.def_path_hash(def_id_1);
```
Comment (nikomatsakis) [inline src/librustc/dep_graph/dep_node.rs]: it'd be good to add a comment clarifying that this is a micro-opt, and the generic impl also suffices

**Review by nikomatsakis — CHANGES_REQUESTED** [review]

**Review by nikomatsakis — APPROVED** [review]

**Comment by michaelwoerister** [comment]
@bors retry

(looks like a time-out)

**Comment by Mark-Simulacrum** [comment]
@bors retry
 - timeout

**Comment by alexcrichton** [comment]
Hm so the [timed out build](https://travis-ci.org/rust-lang/rust/jobs/242913446) compared to the [previous successful build](https://travis-ci.org/rust-lang/rust/jobs/242787492) shows that  stage0-rustc jumped from 1000s to 2600s and stage1-rustc jumped from 1300s to 3500s.

This PR doesn't seem particularly at fault, I'm messaging Travis to hopefully clarify

**Pull request merged**